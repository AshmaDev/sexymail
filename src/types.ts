export interface Config {
  logo?: string;
  colors: {
    headingText: string;
    primaryText: string;
    background: string;
    foreground: string;
    secondary: string;
    primary: string;
    text: string;
  };
  footer?: string;
}

export interface ListItem {
  img: string;
  name: string;
  quantity?: string;
  price?: string;
  features?: string;
}

export interface SummaryItem {
  name: string;
  value: string;
}

export interface Link {
  text: string;
  url: string
};

export interface List {
  items?: ListItem[];
  summary?: SummaryItem[]
}

export interface Seed {
  type: "header" | "text" | "link" | "code" | "list" | "blank";
  value: string & Link & List;
}