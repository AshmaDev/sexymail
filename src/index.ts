import {
  Config,
  Seed,
  SeedValue,
  ListItem,
  SummaryItem,
  Link,
  List,
} from "./types";

class SexyMail {
  private config: Config;

  constructor(config: Config) {
    if (!config) throw new Error("Config is missing");
    this.config = config;
  }

  private createContainer(content: string, info?: string): string {
    return `<div leftmargin="0" marginwidth="0" topmargin="0" marginheight="0" offset="0" style="background-color:${
      this.config.colors.background
    }; height:auto !important;width:100% !important; font-family: Helvetica,Arial,sans-serif !important; padding-top: 40px; margin-bottom: 40px;"> 
      <center>
        <table bgcolor="${
          this.config.colors.foreground
        }" border="0" cellpadding="0" cellspacing="0" style="max-width:600px; width: 100%; background-color:${
      this.config.colors.foreground
    };border:1px solid #e4e2e2;border-collapse:separate !important; border-radius:15px;border-spacing:0;color:#242128; margin:0;padding:40px;" heigth="auto">
          <tbody>
          ${
            this.config.logo
              ? `
            <tr>
              <td align="left" valign="center" style="padding-bottom:40px;border-top:0;height:100% !important;width:100% !important;">
                <center><img src="${this.config.logo}" height="30" style="max-height:30px" /></center>
              </td>
            </tr>`
              : ""
          }
            <tr>
              <td style="padding-top:10px;border-top:1px solid #e4e2e2">${content}</td>
            </tr>
          </tbody>
        </table>
        ${
          info
            ? `
          <table style="margin-top:30px; padding-bottom:20px;; margin-bottom: 40px;">
            <tbody>
              <tr>
                <td align="center" valign="center">
                  <p style="font-size: 12px; text-decoration: none;line-height: 1; color:${this.config.colors.text}; margin-top:0px; margin-bottom:5px; ">${info}</p>
                </td>
              </tr>
            </tbody>
          </table>`
            : ""
        }
      </center>
    </div>`;
  }

  private createHeader(title: string): string {
    return `<h3 style="color:${this.config.colors.headingText}; font-size:18px; line-height: 1.6; font-weight:500;">${title}</h3>`;
  }

  private createText(text: string): string {
    return `<p style="color:${this.config.colors.text}; font-size: 14px; padding-bottom: 10px; line-height: 1.4;">${text}</p>`;
  }

  private createLink(text: string, url: string): string {
    return `<table border="0" cellpadding="0" cellspacing="0" width="100%" style="min-width:100%;border-collapse:collapse;">
        <tbody>
          <tr>
            <td style="padding:15px 0px;" valign="top" align="center"> 
              <table border="0" cellpadding="0" cellspacing="0" style="border-collapse:separate !important;"> 
                <tbody> 
                  <tr> 
                    <td align="center" valign="middle" style="padding:14px;">
                      <a href="${url}" title="${text}" target="_blank" style="font-size: 14px; line-height: 1.5; font-weight: 700; letter-spacing: 1px; padding: 15px 40px; text-align:center; text-decoration:none; color:${this.config.colors.primaryText}; border-radius: 10px; background-color:${this.config.colors.primary};">${text}</a> 
                    </td>
                  </tr>
                </tbody>
              </table>
            </td>
          </tr>
        </tbody>
      </table>`;
  }

  private createCodeBlock(text: string): string {
    return `<p style="background-color:${this.config.colors.background}; color: ${this.config.colors.headingText}; padding: 8px 15px; border-radius: 10px; display: inline-block; margin-bottom:20px; font-size: 24px; line-height: 1.4; font-family: Courier New, Courier, monospace; font-weight: 600; margin-top:0">${text}</p>`;
  }

  private createListItem(item: ListItem): string {
    return `<tr>
        <td style="padding-top:0px; padding-bottom:20px; width:140px "> 
          <img src="${item?.img}" style="width: 114px; height: 85px; object-fit: cover; border-radius: 5px; " /> 
        </td>
        <td style="padding-top:0px; padding-bottom:20px;"> 
          <h4 style="font-size: 16px; line-height: 1; margin-bottom:20px;">
            <a href="#" style="text-decoration: none; color: ${this.config.colors.headingText}; font-weight:500;">${item.name}</a>
          </h4> 
          <p style="font-size: 12px; text-decoration: none; line-height: 1; color:${this.config.colors.text}; margin-top:0px; margin-bottom:0;">${item?.quantity}</p>
          <p style="font-size: 12px; line-height: 1; color:${this.config.colors.text}; margin-top:5px;">${item?.features}</p> 
        </td>
        <td style="padding-top:0px; padding-bottom:20px; text-align: right;"> 
          <p style="font-size: 14px; line-height: 1; color: ${this.config.colors.primary}; margin-top:5px; vertical-align:top; white-space:nowrap;">${item?.price}</p> 
        </td> 
      </tr>`;
  }

  private createSummaryItem(item: SummaryItem): string {
    return `<tr>
        <td colSpan="2" style="padding-top:0px; padding-bottom:5px; text-align: right; color: ${this.config.colors.text};"> 
          <p style="font-size: 12px; line-height: 1; margin-top:5px; vertical-align:top; margin-bottom: 0;">
            <strong>${item?.name}</strong>
          </p>
        </td>
        <td style="padding-top:0px; padding-bottom:5px; text-align: right; padding-left: 15px;">
          <p style="font-size: 14px; line-height: 1; margin-top:5px; vertical-align:top;color:${this.config.colors.text}; white-space:nowrap; margin-bottom: 0;">
            <strong>${item?.value}</strong>
          </p>
        </td>
      </tr>`;
  }

  private createListContainer(items: string): string {
    return `<table style="width: 100%;">${items}</table>`;
  }

  private generateList(obj: {
    items?: ListItem[];
    summary?: SummaryItem[];
  }): string {
    let list = "";

    if (obj.items && obj.items.length > 0) {
      obj.items.forEach((item) => {
        list += this.createListItem(item);
      });
    }

    if (obj.summary && obj.summary.length > 0) {
      obj.summary.forEach((item) => {
        list += this.createSummaryItem(item);
      });
    }

    return this.createListContainer(list);
  }

  public generate(seed: Seed[]): string {
    const elements = {
      header: (text: string) => this.createHeader(text),
      text: (text: string) => this.createText(text),
      link: ({ text, url }: Link) => this.createLink(text, url),
      code: (text: string) => this.createCodeBlock(text),
      list: (obj: List) => this.generateList(obj),
      blank: (content: string) => content,
    };

    const content = seed.map((element: Seed) =>
      elements[element.type](element.value as SeedValue)
    );

    return this.createContainer(content.join(""), this.config.footer);
  }
}

export { SexyMail, Config, Seed };
