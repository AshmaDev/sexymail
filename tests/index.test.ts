import { SexyMail, Config, Seed } from "../src";

describe("SexyMail", () => {
  const config: Config = {
    logo: "http://example.com/logo.png",
    footer: "Sample Footer",
    colors: {
      background: "#ffffff",
      foreground: "#f4f4f4",
      text: "#333333",
      headingText: "#111111",
      primary: "#007bff",
      primaryText: "#ffffff",
      secondary: "#007bff",
    },
  };

  let sexyMail: SexyMail;

  beforeEach(() => {
    sexyMail = new SexyMail(config);
  });

  test("throws error if no config", () => {
    expect(() => new SexyMail(null as any)).toThrow("Config is missing");
  });

  test("generates the correct email HTML", () => {
    const seed: Seed[] = [
      { type: "header", value: "Welcome!" },
      { type: "text", value: "This is a sample email." },
      {
        type: "link",
        value: { text: "Click here", url: "http://example.com" },
      },
    ];

    const emailHTML = sexyMail.generate(seed);

    expect(emailHTML).toMatchSnapshot();
  });
});
