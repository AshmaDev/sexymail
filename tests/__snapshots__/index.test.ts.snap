// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SexyMail generates the correct email HTML 1`] = `
"
      <div style=\\"background-color:#ffffff; height:auto !important;width:100% !important; font-family: Helvetica,Arial,sans-serif !important; padding-top: 40px; margin-bottom: 40px;\\">
        <center>
          <table border=\\"0\\" cellpadding=\\"0\\" cellspacing=\\"0\\" style=\\"max-width:600px; width: 100%; background-color:#f4f4f4;border:1px solid #e4e2e2;border-collapse:separate !important; border-radius:15px;border-spacing:0;color:#242128; margin:0;padding:40px;\\">
            <tbody>
              <tr><td align=\\"left\\" valign=\\"center\\" style=\\"padding-bottom:40px;border-top:0;height:100% !important;width:100% !important;\\"><center><img src=\\"http://example.com/logo.png\\" height=\\"30\\" style=\\"max-height:30px\\" /></center></td></tr>
              <tr><td style=\\"padding-top:10px;border-top:1px solid #e4e2e2\\"><h3 style=\\"color:#111111; font-size:18px; line-height: 1.6; font-weight:500;\\">Welcome!</h3><p style=\\"color:#333333; font-size: 14px; padding-bottom: 10px; line-height: 1.4;\\">This is a sample email.</p>
      <table border=\\"0\\" cellpadding=\\"0\\" cellspacing=\\"0\\" width=\\"100%\\" style=\\"min-width:100%;border-collapse:collapse;\\">
        <tbody>
          <tr>
            <td style=\\"padding:15px 0px;\\" valign=\\"top\\" align=\\"center\\">
              <table border=\\"0\\" cellpadding=\\"0\\" cellspacing=\\"0\\" style=\\"border-collapse:separate !important;\\">
                <tbody>
                  <tr>
                    <td align=\\"center\\" valign=\\"middle\\" style=\\"padding:14px;\\">
                      <a href=\\"http://example.com\\" title=\\"Click here\\" target=\\"_blank\\" style=\\"font-size: 14px; line-height: 1.5; font-weight: 700; letter-spacing: 1px; padding: 15px 40px; text-align:center; text-decoration:none; color:#ffffff; border-radius: 10px; background-color:#007bff;\\">
                        Click here
                      </a>
                    </td>
                  </tr>
                </tbody>
              </table>
            </td>
          </tr>
        </tbody>
      </table></td></tr>
            </tbody>
          </table>
          <table style=\\"margin-top:30px; padding-bottom:20px; margin-bottom: 40px;\\"><tbody><tr><td align=\\"center\\" valign=\\"center\\"><p style=\\"font-size: 12px; text-decoration: none;line-height: 1; color:#333333; margin-top:0px; margin-bottom:5px;\\">Sample Footer</p></td></tr></tbody></table>
        </center>
      </div>"
`;
